ir/src/constraints/list_comprehension.rs

```diff
@@ -0,0 +1,376 @@
+        Some(iterable_type) => match iterable_type {
+            Iterable::Identifier(ident) => {
+                let ident_type = symbol_table.get_type(ident.name())?;
+                match ident_type {
+                    IdentifierType::TraceColumns(size) => {
+                        validate_access(i, size.size())?;
+                        Ok(Expression::NamedTraceAccess(NamedTraceAccess::new(
+                            ident.clone(),
+                            i,
+                            named_trace_access.row_offset(),
+                        )))
+                    }
+                    _ => Err(SemanticError::InvalidListComprehension(
+                        "Iterable should be a trace column".to_string(),
+                    ))?,
+                }
+            }
+            Iterable::Range(_) => Err(SemanticError::InvalidListComprehension(
+                "Iterable cannot be of range type here".to_string(),
+            )),
+            Iterable::Slice(ident, range) => {
+                let ident_type = symbol_table.get_type(ident.name())?;
+                match ident_type {
+                    IdentifierType::TraceColumns(trace_columns) => {
+                        validate_access(i, trace_columns.size())?;
+                        Ok(Expression::NamedTraceAccess(NamedTraceAccess::new(
+                            ident.clone(),
+                            range.start() + i,
+                            named_trace_access.row_offset(),
+                        )))
+                    }
+                    _ => Err(SemanticError::InvalidListComprehension(
+                        "Iterable should be a trace column".to_string(),
+                    ))?,
+                }
+            }
+        },
+    }
+}
+
+/// Parses a list folding expression inside a list comprehension expression.
+fn parse_list_folding(
+    lf_type: &ListFoldingType,
+    expression: &Expression,
+    symbol_table: &SymbolTable,
+    i: usize,
+) -> Result<Expression, SemanticError> {
+    match lf_type {
+        ListFoldingType::Sum(lc) | ListFoldingType::Prod(lc) => {
+            let iterable_context = build_iterable_context(lc)?;
+            let list = unfold_lc(lc, symbol_table)?;
+            let mut acc = parse_lc_expr(expression, &iterable_context, symbol_table, i)?;
+            for elem in list.iter().skip(1) {
+                let expr = parse_lc_expr(elem, &iterable_context, symbol_table, i)?;
+                acc = match lf_type {
+                    ListFoldingType::Sum(_) => Expression::Add(Box::new(acc), Box::new(expr)),
+                    ListFoldingType::Prod(_) => Expression::Mul(Box::new(acc), Box::new(expr)),
+                };
+            }
+            Ok(acc)
+        }
+    }
+}
+
+/// Validates and returns the length of a list comprehension. Checks that the length of all iterables
+/// in the list comprehension is the same and returns an error if it's not.
```

Comment:
nit: the error should be separated into an `# Errors` section. There are several other places with the same problem, so I won't comment them anymore, but please check the rest of the doc comments

Reply:
Makes sense. I generally take the approach of having a separate section if there are more than one error cases but I don't think I've been consistent with that either. I'll add an errors section everywhere.